File: Python3/1 - 200/199 Binary Tree Right Side View.py
# 40ms, 50.62%; 12.8MB, 100.00%
class Solution:
    def rightSideView(self, root: TreeNode) -> list:
        res = []
        self.__helper(root, 1, res)
        return res

    def __helper(self, root: TreeNode, level: int, res: list):
        if not root: return
        if level > len(res): res.append(root.val)
        self.__helper(root.right, level + 1, res)
        self.__helper(root.left, level + 1, res)

File: 1 - 200/199 Binary Tree Right Side View.cc
// 4ms, 73.36%; 9.6MB, 75.68%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> res;
        helper(root, 1, res);
        return res;
    }

private:
    void helper(TreeNode *root, int level, vector<int> &res) {
        if (!root) return;
        if (level > res.size()) res.push_back(root->val);
        helper(root->right, level + 1, res);
        helper(root->left, level + 1, res);
    }
};
